indexing-service/src/main/java/org/apache/druid/indexing/overlord/sampler/InputSourceSampler.java

```diff
@@ -111,38 +113,49 @@ public SamplerResponse sample(
     try (final CloseableIterator<InputRowListPlusRawValues> iterator = reader.sample();
          final IncrementalIndex<Aggregator> index = buildIncrementalIndex(nonNullSamplerConfig, nonNullDataSchema);
          final Closer closer1 = closer) {
-      SamplerResponseRow[] responseRows = new SamplerResponseRow[nonNullSamplerConfig.getNumRows()];
-      int counter = 0, numRowsIndexed = 0;
-
-      while (counter < responseRows.length && iterator.hasNext()) {
-        Map<String, Object> rawColumns = null;
-        try {
-          final InputRowListPlusRawValues inputRowListPlusRawValues = iterator.next();
-
-          if (inputRowListPlusRawValues.getRawValues() != null) {
-            rawColumns = inputRowListPlusRawValues.getRawValues();
-          }
-
-          if (inputRowListPlusRawValues.getParseException() != null) {
-            throw inputRowListPlusRawValues.getParseException();
+      List<SamplerResponseRow> responseRows = new ArrayList<>(nonNullSamplerConfig.getNumRows());
+      int numRowsIndexed = 0;
+
+      while (responseRows.size() < nonNullSamplerConfig.getNumRows() && iterator.hasNext()) {
+        final InputRowListPlusRawValues inputRowListPlusRawValues = iterator.next();
+
+        final List<Map<String, Object>> rawColumnsList = inputRowListPlusRawValues.getRawValuesList();
+
+        final ParseException parseException = inputRowListPlusRawValues.getParseException();
+        if (parseException != null) {
+          if (rawColumnsList != null) {
+            // add all rows to response
+            responseRows.addAll(rawColumnsList.stream()
+                                              .map(rawColumns -> new SamplerResponseRow(rawColumns, null, true, parseException.getMessage()))
+                                              .collect(Collectors.toList()));
```

Comment:
nit: with this change, when a parseException is thrown while parsing a list of rows in one message, those rows in the message will be added as separate rows in `responseRows` but with the same parseException. This can be confusing since the parseException error message seem irrelevant to the associated `rawColumns` in the same `responseRow`. IMO, the better fix in this case would be storing the whole `rawColumnsList` in one `SamplerResponseRow`. Then, the parseException can indicate that it was thrown while parsing one of the rows in `rawColumnsList`. However, this requires a change on the web console side as well. I'm OK with fixing this in a follow-up PR.

Reply:
Yes, returning the whole `rawColumnsList` involves some changes at the web-console side which is currently out of my ability. Thanks for understanding.